### DISCUSSION TITLE
3 workers inserting data in the database, unique constraint error because of race condition?

### DISCUSSION THREAD

**Discussion by notflip** [prisma/prisma#20282] [Q&A]
I have 3 workers that are parsing an input text file, and writing to my database (to a table including a relation). I saw an error saying:
`Unique constraint failed on the constraint: 'dbo.Task'`, and nothing more, when I removed 2 of the workers the issue is fixed.
**Is there any way to make this work using multiple workers?** Thank you.

**Comment by ludralph** [comment]
Hi @notflip 👋 

You may want to ensure workers do not process duplicate data. If the workers are processing the same input text file concurrently, make sure they are assigned different portions of the file to parse. This way, each worker operates on a distinct subset of the data, reducing the likelihood of duplicates. 

If it's absolutely necessary to allow concurrent access to the database by multiple workers, you can introduce a locking mechanism to ensure exclusive access to certain operations. For example, you can use a distributed lock to ensure that only one worker can perform write operations on the specific table that has the unique constraint at a time.

**Reply by janpio** [reply]
Or you could just handle the constraint error if it is expected (as multiple workers might be adding the same data at the same time).

**Reply by notflip** [reply]
That's not a bad idea, so retry the operation using a different related id if a constraint error is returned. I'm going to look into this, thanks.
